### PR TITLE
PSS messenger dapp

### PR DESCRIPTION
**Pull request by aquiladev**
#24 #25

### CODE REVIEW

**Comment by homotopycolimit** [comment]
@zelig, @nolash, can you take a look at this?

**Comment by nolash** [comment]
@aquiladev writing to let you know that I've registered the submission, but haven't so far had the bandwith to go through it. I hope to get back to you early next week. In the meantime, thank you very much for your contribution.

**Comment by nolash** [comment]
@aquiladev I've had a look through the submission now, and I have a couple of questions:

* You seem to have chosen a client-server model here. What do you need the server for? 
* Can I customize the host/port for the websocket when I start the application?

Thanks.

**Comment by aquiladev** [comment]
@nolash 
* The dapp communicates with up and running swarm client, without intermediaries, so you can treat swarm as a server. The only difference between the dapp and others is that they use relative addresses  (at least what I see). In the dapp I need to use websocket and I use Erebos library which communicates with swarm. The library needs urls for connection to http and websocket endpoints of swarm
* There is no setting for websocket endpoint yet

**Comment by nolash** [comment]
@aquiladev ok I understand, thanks.

I gave it a spin with @justelad and we were able to add each other and exchange messages. So far so good. Some questions regarding the pss implementation:

* adding contact is pubkey only, no swarm address. is the address handled in the contact request/response?
* what happens if the internet connection goes down temporarily. Does the app rebuild the websocket connection?
* could you please display the full pubkey in the brower app (or even better, display first and last bytes, with a copy-full-key-hex--to-clipboard mechanism?)
* would you consider adding websocket endpoint?

I'm not qualified to do a proper review of a react app, so @justelad will be taking care of that in the near future.

THanks.

**Comment by aquiladev** [comment]
Hi @nolash,
1. I suppose Erebos library handles it
2. In case of connection lose it will not be restored automatically, the dapp should be refreshed in order to receive/send messages
3. I will implement formatting (display first and last bytes) with copy-to-clipboard
4. Where do you think it would be better to set ws endpoint? I can show popup with ws endpoint field on dapp's start

**Comment by nolash** [comment]
Regarding 1. no, it can't because it has no way of knowing. Either the address must be explicitly given by the user, _or_ it would have to be part of the request/response handshake.

You can specify anything from 0 to 32 bytes of address. The more bytes you specify, the more efficient the routing will be. Right now it's in effect 0 bytes, which means it sends all messages to all peers (along the lines of the basic strategy of whisper). The main point of pss, however, is to provide message _routing_. So could we do either one of the two? THanks.

**Comment by aquiladev** [comment]
1. Not sure that I fully understood it, I will spend time to investigate it
    Current implementation:
    1) on app start I retrieve base address by ```pss_baseAddr``` and public key by ```pss_getPublicKey```
    2) send contact request:
        - set peer ```pss_setPeerPublicKey {receiverPublicKey} {topic}```
        - send message ```pss_sendAsym {receiverPublicKey} {topic} {message}```, message contains base address of sender
    3) accept request by contact:
        - set peer ```pss_setPeerPublicKey {senderPublicKey} {topic}```
        - send message ```pss_sendAsym {senderPublicKey} {topic} {message}```, message contains base address  of receiver
    4) exchange messages ```pss_sendAsym {publicKey} {topic} {message}```

**Would you please point me to the place regarding the address?**

2. Agree
3. Done
4. Done (Pushed changes)

**Comment by nolash** [comment]
> send message pss_sendAsym {senderPublicKey} {topic} {message}, message contains base address of receiver

@aquiladev aha so it _is_ handled in the contact request/response then? The full address of the requester is set with `setPeerPublicKey` on the receiver, and the requester updates from the address in the response? If it's done like this it's all fine.

... I mean - ideally sliding scale "luminosity," of course, but that's for later :)

**Comment by nolash** [comment]
_HOWEVER_ - I see in your calls to `setPeerPublicKey` in `/examples/messenger/src/base/messenger.js` the address argument is omitted. That translates to an zero-byte address. Even if your application message struct contains the address, there is no software that automagically sets this for you.

File: examples/messenger/src/base/messenger.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    this._subscriptions.map(s => s.unsubscribe());
+  }
+
+  async sendContactRequest(key) {
+    const [contactTopic, sharedTopic] = await Promise.all([
+      this.client.pss.stringToTopic(key),
+      this.client.pss.stringToTopic(createRandomString()),
+    ]);
+
+    await this.client.pss.setPeerPublicKey(key, contactTopic);
```
Comment (nolash) [inline examples/messenger/src/base/messenger.js]: Please also supply the address argument to `setPeerPublicKey`. 

The requesting node needs to call this again _with_ the address when response is received.

**Comment by nolash** [comment]
Oh and I forgot one more thing:

Could you please remove the leading `/` in the resource paths? Or are they necessary? Without them the app can be served from a subfolder.

**Comment by aquiladev** [comment]
@nolash there is no routing, ```/``` is not needed and not used

**Comment by nolash** [comment]
If I put the app on my webserver at `localhost/messenger` lets say, I get a bunch of 404s when trying to load.

**Comment by nolash** [comment]
@aquiladev the address stuff seems to work. Good! Curious: If I first send 0x as the address, it gets updated on the response, right?

**Comment by nolash** [comment]
The websocket popup, if I give an alternate value (port 8547 for example) and press enter, the popup just pops back, if I press the button, it works.

**Comment by aquiladev** [comment]
@nolash 
- I've figured out issue with routing, it should be fixed. In order to deploy the dapp you need build (```npm run build```), then copy artifact to target location
- In case of empty address the request will be broadcasted. When receiver accept request, requester will receive and create a peer with address of receiver.
- I fixed issue with ```enter```

**Comment by nolash** [comment]
Great! That's all from me. Thanks :)

**Comment by aquiladev** [comment]
@justelad I agree that the solution design may look not good, can be heavily improved, but this is example dapp, I think it is not needed here

File: examples/messenger/src/App.jsx [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        chats: this.state.chats
+      });
+    });
+  }
+
+  componentDidMount() {
+    this.init()
+  }
+
+  componentWillUnmount() {
```
[the remark this answers is not in the document]
Comment (nolash) [reply examples/messenger/src/App.jsx]: @justelad good question, I don't know, actually.

[the remark this answers is not in the document]
Comment (justelad) [reply examples/messenger/src/App.jsx]: 👍

**Pull request merged**